crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+                    // proportional to that module's freelist
+                    // length. Or in other words, every *slot* should
+                    // be equally likely to be stolen. The
+                    // alternative, where we pick the victim module
+                    // freelist first, means that either a module with
+                    // an affinity freelist of one slot has the same
+                    // chances of losing that slot as one with a
+                    // hundred slots; or else we need a weighted
+                    // random choice among modules, which is just as
+                    // complex as this process.
+                    //
+                    // We don't bother picking an empty slot (no
+                    // established affinity) before a random slot,
+                    // because this is more complex, and in the steady
+                    // state, all slots will see at least one
+                    // instantiation very quickly, so there will never
+                    // (past an initial phase) be a slot with no
+                    // affinity.
+                    let free_list_index = rand::thread_rng().gen_range(0..free_list.len());
+                    let new_id = free_list[free_list_index];
+                    // Remove from both the global freelist and
+                    // per-module freelist, if any.
+                    Self::remove_free_list_item(slot_state, free_list, new_id);
+                    if let &SlotState::Affinity { module, .. } = &slot_state[new_id] {
+                        Self::remove_module_free_list_item(slot_state, per_module, module, new_id);
+                    }
+                    slot_state[new_id] = SlotState::Taken;
+
+                    new_id
+                }
+            }
+        }
+    }
+
+    pub(crate) fn free(&mut self, index: usize, id: Option<CompiledModuleId>) {
+        match self {
+            &mut PoolingAllocationState::NextAvailable(ref mut free_list)
+            | &mut PoolingAllocationState::Random(ref mut free_list) => {
+                free_list.push(index);
+            }
+            &mut PoolingAllocationState::ReuseAffinity {
+                ref mut per_module,
+                ref mut free_list,
+                ref mut slot_state,
+            } => {
+                let free_list_index = free_list.len();
+                free_list.push(index);
+                if let Some(id) = id {
+                    let per_module_list = per_module.entry(id).or_insert_with(|| vec![]);
+                    let per_module_index = per_module_list.len();
+                    per_module_list.push(index);
+                    slot_state[index] = SlotState::Affinity {
+                        module: id,
+                        free_list_index,
+                        per_module_index,
+                    };
+                } else {
+                    slot_state[index] = SlotState::Empty { free_list_index };
```

Comment:
Yeah so calling `free` without an `id` will create new `SlotState::Empty`s which goes against that steady-state comment above (ignoring a global analysis of all `free` callers). So yeah I think removing the `id` param would be nice if we can, since it also helps make it easier to verify statements about the correctness of this system.

Reply:
Indeed. The intent was to be a little more general (one could in theory choose *not* to retain the state that makes a slot have affinity) but as you say that complexifies things a bit and we should probably approach that as we need it. (And it need not necessarily touch the affinity state either; there's little harm in a stray affinity for a module that will never exist again.)